src/Data/Generic/Rep/Enum.purs

```diff
@@ -0,0 +1,72 @@
+module Data.Generic.Rep.Enum
+  ( class GenericEnum
+  , genericSucc
+  , genericPred
+  , genericFromEnum
+  , genericToEnum
+  , genericCardinality
+  , succ'
+  , pred'
+  , construct
+  , fromEnum'
+  , toEnum'
+  , cardinality'
+  ) where
+
+import Data.Generic.Rep
+import Data.Maybe (Maybe(..))
+import Data.Enum (Cardinality(Cardinality))
+import Data.Newtype (unwrap)
+import Prelude ((<$>), (+), (-))
+
+class GenericEnum a where
+  succ' :: a -> Maybe a
+  pred' :: a -> Maybe a
+  fromEnum' :: a -> Int
+  toEnum' :: Int -> Maybe a
+  cardinality' :: Cardinality a
+  construct :: a
```

Comment:
Something doesn't seem right about `construct` to me. I think you need two functions, one for the minimum value and one for the maximum. That way, `succ` and `pred` will be symmetric in the `Sum` case as well.